### PR TITLE
Add Protodef file , Exceptions & authkey

### PR DESCRIPTION
**Pull request by mossplix**

### CODE REVIEW

**Comment by hamiltop** [comment]
Whoa. Lot of code here.

Can we take a step back and describe what we are trying to accomplish with this code?

**Comment by mossplix** [comment]
Ok I am going to annotate it :)

**Comment by hamiltop** [comment]
The auth key and the exceptions look great.

As commented, I'm not super excited about including the compile proto file in the repo. I am, however, very open to the idea of having a single place where we translate from symbols to protocol numbers. I just don't want the fact that the spec is a protobuf to leak into the rest of the code.

One positive aspect of using the proto file is that future upgrades will be easier. If we want to go down this route, we can do a bunch of compile time stuff with macros to produce the desired code.

Can we break the auth and exception stuff into a new pull request (or pull the proto out of this one?) That can be merged then we can discuss the protobuf stuff separately.

File: lib/exrethinkdb/connection.ex [hunk 1]
```diff
@@ -8,29 +19,38 @@ defmodule Exrethinkdb.Connection do
...
   end
 
-  defp handshake(socket) do
-    :ok = :gen_tcp.send(socket, << 0x400c2d20 :: little-size(32) >>)
-    :ok = :gen_tcp.send(socket, << 0 :: little-size(32) >>)
-    :ok = :gen_tcp.send(socket, << 0x7e6970c7 :: little-size(32) >>)
-    {:ok, "SUCCESS" <> << 0 :: size(8)  >>} = :gen_tcp.recv(socket, 8)
-    :ok
+  defp handshake(socket,auth_key) do
+        :ok = :gen_tcp.send(socket, :binary.encode_unsigned(@proto_version,:little))
```
Comment (hamiltop) [inline lib/exrethinkdb/connection.ex]: encode_unsigned doesn't guarantee 4 bytes, which is what the spec wants. Unless I'm misunderstanding it.

**Comment by hamiltop** [comment]
Are we going to go forward with this? My feedback was in the spirit of adjusting the approach, not dismissing it.

**Comment by hamiltop** [comment]
I'm going to close this, but in the very least authkey and exceptions should be added. Feel free to reopen if we want to explore the "registry of opcodes" approach.

**Pull request closed without merging**